packages/lib/src/interactions/SelectToZoom.tsx

```diff
@@ -1,69 +1,76 @@
 import { useThree } from '@react-three/fiber';
 
 import { useVisCanvasContext } from '../vis/shared/VisCanvasProvider';
-import RatioSelectionRect from './RatioSelectionRect';
 import SelectionRect from './SelectionRect';
 import SelectionTool from './SelectionTool';
-import { useMoveCameraTo } from './hooks';
-import type { CommonInteractionProps, Selection } from './models';
-import { getEnclosedRectangle, getRatioRespectingRectangle } from './utils';
+import Box from './box';
+import { useZoomOnBox } from './hooks';
+import type { CommonInteractionProps, Rect3, Selection } from './models';
 
 type Props = CommonInteractionProps;
 
 function SelectToZoom(props: Props) {
-  const { canvasSize, canvasRatio, visRatio, dataToWorld } =
+  const { canvasSize, canvasRatio, visRatio, worldToData } =
     useVisCanvasContext();
 
   const { width, height } = canvasSize;
   const keepRatio = visRatio !== undefined;
 
   const camera = useThree((state) => state.camera);
-  const moveCameraTo = useMoveCameraTo();
+  const zoomOnBox = useZoomOnBox();
 
-  function onSelectionEnd(selection: Selection) {
-    const [worldStart, worldEnd] = getRatioRespectingRectangle(
-      ...selection.data,
-      keepRatio ? canvasRatio : undefined
-    ).map(dataToWorld); // work in world coordinates as we need to act on the world camera
+  function computeZoomBox(worldSelection: Rect3): Box {
+    const { position, scale } = camera;
+    const zoomBox = Box.fromPoints(...worldSelection);
 
-    if (worldStart.x === worldEnd.x || worldStart.y === worldEnd.y) {
-      return;
+    if (!keepRatio) {
+      return zoomBox;
     }
 
-    const zoomRect = getEnclosedRectangle(worldStart, worldEnd);
-    const { center: zoomRectCenter } = zoomRect;
+    const fovBox = Box.empty(position).expandBySize(
+      width * scale.x,
+      height * scale.y
+    );
+
+    return zoomBox.expandToRatio(canvasRatio).keepWithin(fovBox);
+  }
 
-    // Change scale first so that moveCameraTo computes the updated camera bounds
-    camera.scale.set(zoomRect.width / width, zoomRect.height / height, 1);
-    camera.updateMatrixWorld();
+  function onSelectionEnd(selection: Selection) {
+    const [worldStart, worldEnd] = selection.world;
+    if (worldStart.x === worldEnd.x || worldStart.y === worldEnd.y) {
+      return;
+    }
 
-    moveCameraTo(zoomRectCenter);
+    zoomOnBox(computeZoomBox(selection.world));
```

Comment:
Note that we didn't use to bother calling `clampRectangleToVis` before `moveCameraTo`, since `moveCameraTo` already takes care of keeping the FOV within the visualization area.

Now, `computeZoomBox` always calls `keepWithin`. I think the explicitness and consistency between `onSelectionEnd` and the render prop is good. It might even allow us to remove the call to `moveCameraTo` from `zoomOnBox` and to just set the camera position directly to the box' center.